### DISCUSSION TITLE
How to create a dynamic filter of type `IObservable<Func<Data,bool>>`?

### DISCUSSION THREAD

**Discussion by Blazero** [reactivemarbles/DynamicData#754] [Q&A]
In the [README.md](https://github.com/reactivemarbles/DynamicData#filtering) of this project, there is an example to filter a change set dynamically
```
IObservable<Func<Person,bool>> observablePredicate = ...;
var myFilteredObservable = myPeopleObservable.Filter(observablePredicate); 
```
But how can I create the `IObservable<Func<Person,bool>> observablePredicate = ...;` and modify it when needed?

**Comment by Blazero** [comment]
I've read [a old blog](https://dynamic-data.org/2015/04/22/dynamically-sort-filter-and-page-data/). The `var filterController = new FilterController<T>();` in this blog seems to be what I need but now it's removed.

**Comment by Blazero** [comment]
I've read the example [Dynamic Trader](https://github.com/RolandPheasant/Dynamic.Trader/blob/3f0f67b7749865456d4e698815bddb7262b1485d/Trader.Client/Views/PagedDataViewer.cs#L39) and the filter in it is created by
```
var filter = this.WhenValueChanged(t => t.SearchText)
                .Throttle(TimeSpan.FromMilliseconds(250))
                .Select(BuildFilter);
```
But when I try this in my project, the `ItemsSource` property of DataGrid throws an exception of `Call from invalid thread`. I assume the Observer of this filter runs on a backgroud thread but I have no idea how to fix this.
Besides, I don't need to change the filter every time a value is changed.

**Reply by JakenVeina** [reply]
In general, threading in a GUI library amounts to proper placement of `.ScheduleOn(RxApp.MainThreadScheduler)` (or `.ScheduleOn(DispatcherScheduler)` if you're not using ReactiveUI). This should usually be placed before any operations that affect GUI components, so, before a `.Bind()` to an observable collection that the GUI subscribes to, or before a `.Subscribe()` that changes a property upon an `INotifyPropertyChanged` object that the UI subscribes to, or before a `.Do()` that emits side-effects that affect the GUI, etc.,

That doesn't mean you ALWAYS need this, because changes that affect GUI components generally *come from* a GUI component in the first place, and thus are already running on the GUI render thread. In this case, `.Throttle()` is the source of activity that doesn't start from the GUI thread, because it doesn't come from a GUI component it comes from a behind-the-scenes timer that's measuring time delays of 250 milliseconds.

You can either drop `.ScheduleOn()` before the `.Bind()` call, within the stream that's using this throttled filter, or you can supply the proper scheduler to `.Throttle()` itself. Since it's inherently a timing operator, it accepts an `IScheduler` parameter directly.
